search-indexer/src/orchestrator/mod.rs

```diff
@@ -0,0 +1,333 @@
+                        (events.saturating_sub(prev_events) as f64) / elapsed_secs
+                    } else {
+                        0.0
+                    };
+
+                    let docs_per_sec = if elapsed_secs > 0.0 {
+                        (docs.saturating_sub(prev_docs) as f64) / elapsed_secs
+                    } else {
+                        0.0
+                    };
+
+                    info!(
+                        events_processed = events,
+                        documents_indexed = docs,
+                        events_per_sec = format!("{:.2}", events_per_sec),
+                        documents_per_sec = format!("{:.2}", docs_per_sec),
+                        "Processing progress"
+                    );
+
+                    // Update previous values for next calculation
+                    prev_events = events;
+                    prev_docs = docs;
+                    prev_time = now;
+                }
+            }
+        }
+
+        // Don't flush remaining documents on shutdown - we can't ACK them to Kafka anyway
+        // since the consumer is shutting down. Any pending documents will be re-consumed
+        // on next startup (at-least-once delivery semantics).
+
+        // Wait for consumer to finish
+        let _ = consumer_handle.await;
+
+        let final_events = self.total_events_processed.load(Ordering::Relaxed);
+        let final_docs = self.total_documents_indexed.load(Ordering::Relaxed);
+        info!(
+            total_events_processed = final_events,
+            total_documents_indexed = final_docs,
+            "Orchestrator shutdown complete"
+        );
+        Ok(())
+    }
+
+    /// Process a batch of events through the ingest.
+    ///
+    /// This method ensures documents are actually indexed in OpenSearch before returning Ok.
+    /// The caller should only ACK to Kafka after this method returns successfully.
+    async fn process_events(
+        &mut self,
+        events: Vec<crate::consumer::EntityEvent>,
+    ) -> Result<(), IngestError> {
+        let event_count = events.len();
+        self.total_events_processed
+            .fetch_add(event_count as u64, Ordering::Relaxed);
+
+        debug!(event_count = event_count, "Processing batch of events");
+
+        // Transform events to documents
+        let processed = self.processor.process_batch(events)?;
+
+        if processed.is_empty() {
+            debug!("No documents to index after processing");
+            return Ok(());
+        }
+
+        // Count documents to be indexed
+        let index_count = processed
+            .iter()
+            .filter(|e| matches!(e, crate::processor::ProcessedEvent::Index(_)))
+            .count();
+
+        // Load into search index and process all operations immediately.
+        // This ensures we only ACK to Kafka after documents are actually indexed.
+        let operation_summaries = self.loader.load(processed).await?;
```

Comment:
i think here we are tightly coupling the processing with the loader, should them run in different threads. 

I imagine the workflow as 3 different tokio tasks with mpsc communication between them